File: 0633_sum_of_square_numbers/judge_square_sum.java
class Solution {
    // Time: O(log(n))
    // Space: O(1)
    public boolean judgeSquareSum(int c) {
        if (c < 0) {
            return false;
        }
        int start = 0, end = (int)Math.sqrt(c);
        while (start <= end) {
            int curr = start * start + end * end;
            if (curr == c) {
                return true;
            }
            if (curr < c) {
                start++;
            } else {
                end--;
            }
        }
        return false;
    }
}

File: 0633_sum_of_square_numbers/judge_square_sum.py
def judge_square_sum(c: int) -> bool:
    """
    Time: O(log(n))
    Space: O(1)
    """
    if c < 0:
        return False
    start, end = 0, int(c**0.5)
    while start <= end:
        curr = start**2 + end**2
        if curr == c:
            return True
        if curr < c:
            start += 1
        else:
            end -= 1
    return False
        